src/App.tsx

```diff
@@ -1,68 +1,74 @@
-import {
-  BrowserRouter as Router,
-  Switch,
-  Route,
-  Link
-} from 'react-router-dom';
-import { ProductsPage } from './ProductsPage';
-import { ProductPage } from './ProductPage';
-import { API } from './types';
-import {getProductsList, getProductById} from './api';
-import logo from './logo.svg';
-import './App.css';
+import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
+import { ProductsPage } from "./ProductsPage";
+import { ProductPage } from "./ProductPage";
+import { API, CartItem } from "./types";
+import { useCallback, useState } from "react";
+import { getProductsList, getProductById } from "./api";
+import { CartContext } from "./Cart";
+import "./App.css";
+import { Header } from "./Header";
 
-const defaultAPI = {getProductsList, getProductById};
+const defaultAPI = { getProductsList, getProductById };
 
 export default function App(props: AppProps) {
+  const { api = defaultAPI } = props;
+  const [cart, setCart] = useState<CartItem[]>([]);
 
-  const {api = defaultAPI} = props;
+  const addToCart = useCallback((product_id: number, quantity: number) => {
+    setCart((currentCart) => {
+      let indexOfProduct = currentCart.findIndex(
+        (item) => item.product_id === product_id
+      );
+
+      if (indexOfProduct > -1) {
+        return currentCart.map((item, index) => ({
+          ...item,
+          quantity:
+            index === indexOfProduct ? item.quantity + quantity : item.quantity,
+        }));
```

Comment:
Avoid looping over the cart items as you already know the item index:

```suggestion
        return [
            ...currentCart.slice(0, indexOfProduct),
            { product_id, quantity: currentCart[indexOfProduct].quantity + quantity },
            ...currentCart.slice(indexOfProduct + 1)
        ];
```